mcpserver/tools.go

```diff
@@ -75,30 +82,52 @@ func RunStdioServer(ctx context.Context, dag *dagger.Client) error {
 	return stdioSrv.Listen(ctx, os.Stdin, os.Stdout)
 }
 
-var tools = []*Tool{}
+var tools = []MCPHandler{}
 
-func registerTool(tool ...*Tool) {
-	for _, t := range tool {
-		tools = append(tools, wrapTool(t))
-	}
+func registerTool[R Request, V Response](tool *Tool[R, V]) {
+	tools = append(tools, createHandler(tool))
 }
 
-func wrapTool(tool *Tool) *Tool {
-	return &Tool{
+func createHandler[R Request, V Response](tool *Tool[R, V]) MCPHandler {
+	return MCPHandler{
 		Definition: tool.Definition,
 		Handler: func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
 			slog.Info("Tool called", "tool", tool.Definition.Name)
 			defer func() {
 				slog.Info("Tool finished", "tool", tool.Definition.Name)
 			}()
-			return tool.Handler(ctx, request)
+			// FIXME(aluzzardi): Backward compat, remove this
+			if tool.Handler != nil {
```

Comment:
🔧 **Maintainability (Medium):** The backward compatibility comment indicates this should be removed, but there's no clear migration plan. Consider adding a deprecation log or timeframe for removal

```suggestion
			// FIXME(aluzzardi): Backward compat, remove this after migrating all tools
			if tool.Handler != nil {
				slog.Warn("Using deprecated Handler field, please migrate to ActualHandler", "tool", tool.Definition.Name)
				return tool.Handler(ctx, request)
			}
```